### PR TITLE
8306703: JFR: Summary views

### PR DESCRIPTION
**Pull request by egahlin**
Could I have a review of an enhancement that adds a view command to jfr.

Testing: tier1, tier2 + jdk/jdk/jfr

For the change to work properly when streaming, fix of 8307738 needs to be applied.

To simplify the review, changes not relevant to the feature, but that can use classes in jdk.jfr.internal.util package, will be refactored after this change has been integrated. Possibly in JDK 22.

Thanks
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8306704](https://bugs.openjdk.org/browse/JDK-8306704) to be approved

### Issues
 * [JDK-8306703](https://bugs.openjdk.org/browse/JDK-8306703): JFR: Summary views
 * [JDK-8306704](https://bugs.openjdk.org/browse/JDK-8306704): JFR: Summary views (**CSR**)

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**) ⚠️ Review applies to [e4e4bba1](https://git.openjdk.org/jdk/pull/14104/files/e4e4bba14e8603957bafab2a78b3a72835b5a6e7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14104/head:pull/14104` \
`$ git checkout pull/14104`

Update a local copy of the PR: \
`$ git checkout pull/14104` \
`$ git pull https://git.openjdk.org/jdk.git pull/14104/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14104`

View PR using the GUI difftool: \
`$ git pr show -t 14104`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14104.diff">https://git.openjdk.org/jdk/pull/14104.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14104#issuecomment-1559829206)

### CODE REVIEW

File: test/jdk/jdk/jfr/tool/TestView.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private static void testIncorrectUsage() throws Throwable {
+        OutputAnalyzer output = ExecuteHelper.jfr("view");
+        output.shouldContain("missing file");
+
+        output = ExecuteHelper.jfr("view", "missing.jfr");
+        output.shouldContain("could not open file ");
+
+        Path file = Utils.createTempFile("faked-file",  ".jfr");
```
Comment (turbanoff) [inline test/jdk/jdk/jfr/tool/TestView.java:59]:
```suggestion
        Path file = Utils.createTempFile("faked-file", ".jfr");
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        }
+        if (tokenizer.accept("DESC")) {
+            return SortOrder.DESCENDING;
+        }
+        return SortOrder.NONE;
+    }
+
+    private String text() throws ParseException {
+        if (tokenizer.peekChar() != '\'') {
+            throw new ParseException("Expected text to start with a single quote character",  position());
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java:230]:
```suggestion
            throw new ParseException("Expected text to start with a single quote character", position());
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            if (c != '.' && c != '[' && c!= ']' && c != '|' && !valid) {
+                throw new ParseException("Not a valid field name: " + name, position());
+            }
+        }
+        return name;
+    }
+
+    private Property property() throws ParseException {
+        String text = tokenizer.next();
+        Consumer<Field> style =  switch (text.toLowerCase()) {
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java:270]:
```suggestion
        Consumer<Field> style = switch (text.toLowerCase()) {
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            } catch (NumberFormatException nfe) {
+                // Fall through
+            }
+            throw new ParseException("Expected an integer after LIMIT", position());
+        }
+        return Integer.MAX_VALUE;
+    }
+
+    @Override
+    public void close() throws  ParseException {
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryParser.java:333]:
```suggestion
    public void close() throws ParseException {
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        } else {
+            return formatPositiveDuration(roundedDuration);
+        }
+    }
+
+    private static String formatPositiveDuration(Duration d){
+        if (d.compareTo(MICRO_SECOND) < 0) {
+            // 0.000001 ms - 0.000999 ms
+            double outputMs = (double) d.toNanosPart() / 1_000_000;
+            return String.format("%.6f ms",  outputMs);
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java:73]:
```suggestion
            return String.format("%.6f ms", outputMs);
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        } else if (d.compareTo(SECOND) < 0) {
+            // 0.001 ms - 999 ms
+            int valueLength = countLength(d.toNanosPart());
+            int outputDigit = NANO_SIGNIFICANT_FIGURES - valueLength;
+            double outputMs = (double) d.toNanosPart() / 1_000_000;
+            return String.format("%." + outputDigit + "f ms",  outputMs);
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java:79]:
```suggestion
            return String.format("%." + outputDigit + "f ms", outputMs);
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        } else if (d.compareTo(MINUTE) < 0) {
+            // 1.00 s - 59.9 s
+            int valueLength = countLength(d.toSecondsPart());
+            int outputDigit = MILL_SIGNIFICANT_FIGURES - valueLength;
+            double outputSecond = d.toSecondsPart() + (double) d.toMillisPart() / 1_000;
+            return String.format("%." + outputDigit + "f s",  outputSecond);
+        } else if (d.compareTo(HOUR) < 0) {
+            // 1 m 0 s - 59 m 59 s
+            return String.format("%d m %d s",  d.toMinutesPart(), d.toSecondsPart());
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java:88]:
```suggestion
            return String.format("%d m %d s", d.toMinutesPart(), d.toSecondsPart());
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        } else if (d.compareTo(DAY) < 0) {
+            // 1 h 0 m - 23 h 59 m
+            return String.format("%d h %d m",  d.toHoursPart(), d.toMinutesPart());
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java:91]:
```suggestion
            return String.format("%d h %d m", d.toHoursPart(), d.toMinutesPart());
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    private static int countLength(long value){
+        return (int) Math.log10(value) + 1;
+    }
+
+    private static Duration roundDuration(Duration d) {
+        if (d.equals(Duration.ZERO)) {
+            return d;
+        } else if(d.isNegative()){
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/ValueFormatter.java:105]:
```suggestion
        } else if (d.isNegative()){
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/FieldFormatter.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            return format(field, st.getFrames().getFirst(), compact);
+        }
+        if (object instanceof RecordedThreadGroup tg) {
+            return tg.getName();
+        }
+        if (object instanceof RecordedMethod m) {
+            return ValueFormatter.formatMethod(m, compact);
+        }
+        if (object instanceof RecordedClass clazz) {
+            String text =  ValueFormatter.formatClass(clazz);
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/FieldFormatter.java:94]:
```suggestion
            String text = ValueFormatter.formatClass(clazz);
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryPrinter.java [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        }
+        return false;
+    }
+
+    private void printFields(EventType type, List<EventType> allTypes) {
+        out.println();
+        out.println("" + type.getName() + ":");
+        out.println();
+        for (ValueDescriptor f : type.getFields()) {
+            String typeName =  Utils.makeSimpleName(f.getTypeName());
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/QueryPrinter.java:167]:
```suggestion
            String typeName = Utils.makeSimpleName(f.getTypeName());
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/query/Function.java [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.StringJoiner;
+
+abstract class Function {
+
+    abstract public void add(Object value);
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/query/Function.java:38]: let's use blessed modifiers order
```suggestion
    public abstract void add(Object value);
```

**Review by mgronlun — APPROVED** [review]
This is great work Erik! It is going to help folks better understand what is important, and the extension capabilities are numerous. Thank you very much!

**Comment by egahlin** [comment]
> This is great work, Erik! It is going to help folks better focus to understand what is important and the extension capabilities are numerous. Thank you very much!

Thanks for the review! It was a lot to go through.

**Pull request closed without merging**